### PR TITLE
Refactor StatusProjectController

### PR DESCRIPTION
**Pull request by vpereira**
* Break the method `find_relationship_for_packages` in smaller methods
* Add memoization for `package_hash`
* Improve ruby dialect for `role_from_cache`, `user_from_cache`,
  `group_from_cache`

### CODE REVIEW

**Comment by vpereira** [comment]
I tried to switch the caching scheme, and the code was looking code but unfortunately it doesn't work or maybe my benchmark is wrong (or caching and benchmark is wrong, who knows) 

```
require 'ostruct'
require 'prime'

require 'benchmark'


class BugBurgerApplicationController
  def initialize
    @what_a_hash = {}
  end

  def what_a_cache(k)
    @what_a_hash[k] ||= calculate_the_secret_of_universe[k]
  end

  def calculate_the_secret_of_universe
    # return 42
    Hash[my_structs.map {|x| [x.foo.to_sym, Prime.prime_division(x.bar) ] } ]
  end

  def package_from_cache(key, package_hash)
      @pcache ||= Hash.new do |h, k|
        h[k] = package_hash[k]
      end
      @pcache[key]
  end
  private
  def my_structs 
    (1..100_103).step(2).collect do |e| 
      OpenStruct.new(foo: "key_#{e}", bar: rand(1024..31337)) 
    end
  end

end


if $0 == __FILE__

  # top secret new startup
  bb = BugBurgerApplicationController.new

  Benchmark.bm(38) do |b|
    b.report("calculate the secret of the universe") { bb.calculate_the_secret_of_universe[:key_10] }
    b.report("warmup the new cache                ") { bb.package_from_cache(:key_15, bb.calculate_the_secret_of_universe) }
    b.report("use the new cache                   ") { bb.package_from_cache(:key_15, bb.calculate_the_secret_of_universe) }
    b.report("warmup the old cache                ") { bb.what_a_cache(:key_33) }
    b.report("use the old cache                   ") { bb.what_a_cache(:key_33) }
  end
end

```
running it:

```
vpereira@kimura:~> ruby 1.rb 
                                             user     system      total        real
calculate the secret of the universe     0.917438   0.060119   0.977557 (  0.977893)
warmup the new cache                     0.949821   0.003869   0.953690 (  0.953751)
use the new cache                        0.974520   0.000056   0.974576 (  0.974953)
warmup the old cache                     1.043386   0.000000   1.043386 (  1.043441)
use the old cache                        0.000004   0.000000   0.000004 (  0.000003)
```

so the old cache (how its now implemented) was the only one working :P

File: src/api/app/controllers/status_project_controller.rb [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
   def show
     dbproj = Project.get_by_name(params[:project])
     @packages = ProjectStatus::Calculator.new(dbproj).calc_status
-    find_relationships_for_packages(@packages)
+    find_relationships_for_packages
   end
 
   private
 
+  def initialize_caches
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: I think the more pressing question is. Why do we cache this?

**Comment by vpereira** [comment]
Good catch @coolo! I'm using now a lazy hash creation and now it looks good:

```
require 'ostruct'
require 'prime'

require 'benchmark'


class BugBurgerApplicationController
  def initialize
    @what_a_hash = {}
  end

  def what_a_cache(k)
    @what_a_hash[k] ||= calculate_the_secret_of_universe[k]
  end

  def calculate_the_secret_of_universe
    # return 42
    Hash.new { my_structs.map {|x| [x.foo.to_sym, Prime.prime_division(x.bar) ] } }
  end

  def package_from_cache(key, package_hash)
      @pcache ||= Hash.new do |h, k|
        h[k] = package_hash[k]
      end
      @pcache[key]
  end
  private
  def my_structs 
    (1..100_103).step(2).collect do |e| 
      OpenStruct.new(foo: "key_#{e}", bar: rand(1024..31337)) 
    end
  end
end


if $0 == __FILE__

  # top secret new startup
  bb = BugBurgerApplicationController.new

  Benchmark.bm(38) do |b|
    b.report("calculate the secret of the universe") { bb.calculate_the_secret_of_universe[:key_10] }
    b.report("warmup the new cache                ") { bb.package_from_cache(:key_15, bb.calculate_the_secret_of_universe) }
    b.report("use the new cache                   ") { bb.package_from_cache(:key_15, bb.calculate_the_secret_of_universe) }
    b.report("warmup the old cache                ") { bb.what_a_cache(:key_33) }
    b.report("use the old cache                   ") { bb.what_a_cache(:key_33) }
  end
end
```

```
vpereira@kimura:~> ruby 1.rb 
                                             user     system      total        real
calculate the secret of the universe     1.056629   0.023816   1.080445 (  1.093634)
warmup the new cache                     0.997403   0.043825   1.041228 (  1.054281)
use the new cache                        0.000007   0.000000   0.000007 (  0.000007)
warmup the old cache                     1.260633   0.008030   1.268663 (  1.284038)
use the old cache                        0.000003   0.000000   0.000003 (  0.000005)
vpereira@kimura:~> 
```

File: src/api/app/controllers/status_project_controller.rb (hunk 1 above)
Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: memoizing maybe helps in the ```each`` block?

Comment (hennevogel) [reply src/api/app/controllers/status_project_controller.rb]: maybe?

File: src/api/app/controllers/status_project_controller.rb [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
+    @package_hash = {}
+    @usercache = {}
+    @groupcache = {}
+    @rolecache = {}
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: The last 3 are also unused by now...

File: src/api/app/controllers/status_project_controller.rb [hunk 3]
```diff
@@ -1,12 +1,21 @@
...
+  end
+
   def role_from_cache(role_id)
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: I guess someone did not know about the `ActiveRecord` query cache...

File: src/api/app/controllers/status_project_controller.rb [hunk 4]
```diff
@@ -1,41 +1,68 @@
...
+    find_relationships_for_packages
   end
 
   private
 
+  def initialize_caches
+    @package_hash = {}
+  end
+
   def role_from_cache(role_id)
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: There is an `ActiveRecord` query cache. No need to implement a cache on top of it.

File: src/api/app/controllers/status_project_controller.rb (hunk 1 above)
Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: :musical_note: Hey, I just wrote it and this is craaaazy
But here's my PoC, so call me maaaybe  :musical_note:

File: src/api/app/controllers/status_project_controller.rb [hunk 5]
```diff
@@ -1,41 +1,68 @@
...
 
-  def find_relationships_for_packages(packages)
-    @usercache = {}
-    @groupcache = {}
-    @rolecache = {}
-    package_hash = {}
-    packages.each_value do |p|
-      package_hash[p.package_id] = p
-      package_hash[p.develpack.package_id] = p.develpack if p.develpack
+  def package_hash(packages)
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: We know `find_relationships_for_packages` runs first. We know it will fill `@package_cache`. We know all subsequent calls to `package_hash` will use `@package_cache`. What are we doing here?

File: src/api/app/controllers/status_project_controller.rb [hunk 6]
```diff
@@ -1,41 +1,68 @@
...
-    relationships = Relationship.where(package_id: package_hash.keys).pluck(:package_id, :user_id, :group_id, :role_id)
-    relationships.each do |package_id, user_id, group_id, role_id|
+    @package_hash
+  end
+
+  def relationships_for_package(keys)
+    Relationship.where(package_id: keys).pluck(:package_id, :user_id, :group_id, :role_id)
+  end
+
+  def add_person(user_id, role_id, package_id)
```
Comment (hennevogel) [inline src/api/app/controllers/status_project_controller.rb]: `package_hash(@packages)[package_id]` is just `Package.find(package_id)`
`user_from_cache(user_id)` is just `User.find(user_id)`
`role_from_cache(rold_id)` is just `Role.find(role_id)`

Just another useless cache on top of the `ActiveRecord` query cache IMHO.

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: I think it was discussed here with @coolo https://github.com/openSUSE/open-build-service/pull/5765

@coolo knows (TM)

File: src/api/app/controllers/status_project_controller.rb (hunk 4 above)
Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: @hennevogel I think merging oder don't merging this PR, the caching will be there anyway :)

File: src/api/app/controllers/status_project_controller.rb (hunk 6 above)
Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: I wouldn't remove the caching unless we are 100% that it is useless:

```
vpereira@kimura:~/open-build-service> time osc api '/status/project/openSUSE:Factory' 1>/dev/null

real	2m40.098s
user	0m0.326s
sys	0m0.210s
```
```
vpereira@kimura:~/open-build-service> time osc api '/status/project/openSUSE:Factory' 1>/dev/null

real	0m53.210s
user	0m0.376s
sys	0m0.101s
```

Comment (coolo) [reply src/api/app/controllers/status_project_controller.rb]: the cache we're talking about is memorize - so you can't measure its impact without modifying the code. The cache you just measure is the backend filesystem's cache (yes, it's really that sad)

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: @coolo yes, I know. What I wanted to point out is that as it is, is already bad enough.. so to remove the memoization just if we are 100% sure, that there is no gain with the memoization. I would prefer to don't drop it in my PR.

Comment (coolo) [reply src/api/app/controllers/status_project_controller.rb]: I see

File: src/api/app/controllers/status_project_controller.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/app/controllers/status_project_controller.rb]: > @hennevogel 10 hours ago
> There is an ActiveRecord query cache. No need to implement a cache on top of it.

This sounds [familiar](https://github.com/openSUSE/open-build-service/pull/5765#issuecomment-417647592).

@vpereira And if this is really needed for performance reasons I would check if we could reuse the role caching we have in the [model](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/role.rb#L46).

File: src/api/app/controllers/status_project_controller.rb [hunk 7]
```diff
@@ -1,41 +1,68 @@
...
+      @package_hash[pkg.package_id] = pkg
+      @package_hash[pkg.develpack.package_id] = pkg.develpack if pkg.develpack
     end
-    relationships = Relationship.where(package_id: package_hash.keys).pluck(:package_id, :user_id, :group_id, :role_id)
-    relationships.each do |package_id, user_id, group_id, role_id|
+    @package_hash
+  end
+
+  def relationships_for_package(keys)
+    Relationship.where(package_id: keys).pluck(:package_id, :user_id, :group_id, :role_id)
```
Comment (bgeuken) [inline src/api/app/controllers/status_project_controller.rb]: This looks like a scope

File: src/api/app/controllers/status_project_controller.rb (hunk 6 above)
Comment (bgeuken) [reply src/api/app/controllers/status_project_controller.rb]: Great quote:+1: But what does this actually mean? Do we know by now if and why this needs to be cached? If the route is heavily used we might rather do some [page](https://guides.rubyonrails.org/caching_with_rails.html#page-caching) or [action](https://guides.rubyonrails.org/caching_with_rails.html#action-caching) caching. If the bottleneck is in the backend we should improve it there, etc.

Are we confident that we optimize the right thing? If not I would rather spend some time investigating this instead of spending time refactoring code we might don't need:-)

Comment (coolo) [reply src/api/app/controllers/status_project_controller.rb]: I have no idea if this is really needed any longer - I'd be just very careful with these generic statements about rails doing the right thing. But the route isn't the most often used routed nor do these 3 functions hurt anyone - so I wonder how much effort you want to put into this investigation.

File: src/api/app/controllers/status_project_controller.rb (hunk 4 above)
Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: @bgeuken the code exists already, I just refactored it.  I did not implement the caching there. If you say that I can remove it, I do it..  regarding the usage of  ```Rails.cache.fetch``` vs memoization, I think both are valid, and we have probably both in our code base, Rails had even it's own ``` ActiveSupport::Memoizable``` but then Jose Valim, decided that the "ruby way" was good enough https://github.com/rails/rails/commit/36253916b0b788d6ded56669d37c96ed05c92c5c 

TL;DR:
You can just tell me, please remove that, and I do, I have no strong opinion on that. Otherwise I will keep it as it is (or will research more if needed), either way, IMO it's unrelated with the PR.

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: > This sounds familiar.
> 
> @vpereira And if this is really needed for performance reasons I would check if we could reuse the role caching we have in the model.

@bgeuken I posted it [here](https://github.com/openSUSE/open-build-service/pull/5805/files#r215916132)

File: src/api/app/controllers/status_project_controller.rb (hunk 6 above)
Comment (hennevogel) [reply src/api/app/controllers/status_project_controller.rb]: We generally like to understand WTF is going on and record this for the developers who come after us...

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: Ok, I will test it on monday to see if its used

**Comment by vpereira** [comment]
@bgeuken please look my [benchmark](https://github.com/openSUSE/open-build-service/pull/5805#issuecomment-419381641), it includes the ```ActiveSupport::Cache::MemoryStore```, which is slower than ```memoization```. I think it is a rule of thumb of rails caching: If you want to cache the results of operations to be further used in a loop, in the same instance of a class, and not necessarily useful for other instances or objects, then use memoization otherwise, ```Rails.cache```.. does it make sense?

**Comment by vpereira** [comment]
@hennevogel @bgeuken @coolo 

So I removed the memoization for role, group, and user. There are the results:

### cache warmup

```
Completed 200 OK in 784626ms (Views: 11736.2ms | ActiveRecord: 205427.1ms | Backend: 214892.2ms)
```

### Memoizing roles, users, groups and package_hash:

```
Completed 200 OK in 15328ms (Views: 10316.6ms | ActiveRecord: 160.5ms | Backend: 0.0ms)
Completed 200 OK in 16072ms (Views: 11386.9ms | ActiveRecord: 164.5ms | Backend: 0.0ms)
```

### Memoizing just package_hash (no memoization of roles, users, groups)

```
Completed 200 OK in 26366ms (Views: 10748.4ms | ActiveRecord: 455.4ms | Backend: 0.0ms)
Completed 200 OK in 27440ms (Views: 11362.5ms | ActiveRecord: 466.0ms | Backend: 0.0ms)
```

I create 10_000 packages, and I do make sure that cache is enabled with ```rake dev:cache```. With memoization disabled, I see in the log the cache being hit for roles, users and groups. However as you can see in my benchmark, to access a memory store via ```Rails.cache.fetch``` is more expensive than memoization. Probably with memcache is even more expensive..

The code is asking me here:
So should I stay as it is or should I go?

File: src/api/app/controllers/status_project_controller.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/app/controllers/status_project_controller.rb]: Well, you are refactoring the code around these caching methods as well as the caching code. It's the second time we dance around this code now. I think it's worth to check if we even need this code before refactoring it (and if we cache the right thing when there is some doubt).

> vpereira 3 days ago
> @bgeuken the code exists already, I just refactored it. I did not implement the caching there. If you say that I can remove it, I do it..

You don't need me to answer that question:-)

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: > Well, you are refactoring the code around these caching methods as well as the caching code. It's the second time we dance around this code now.

lambada

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: I'm the last guy @bgeuken

File: src/api/app/controllers/status_project_controller.rb (hunk 6 above)
Comment (bgeuken) [reply src/api/app/controllers/status_project_controller.rb]: > coolo 3 days ago
> I have no idea if this is really needed any longer - I'd be just very careful with these generic statements about rails doing the right thing. 

I am not saying that we should drop it without checking. I am saying that we should have a look and verify that we need it when there is doubt.

> so I wonder how much effort you want to put into this investigation.

It's the second time now within short time that we bump into this. Instead of putting effort into discussing this a third time I would spend an hour and address this.
Beside I find it an odd pattern to act like these are ancient monuments that can not be touched nor understand.

Comment (vpereira) [reply src/api/app/controllers/status_project_controller.rb]: >  this a third time I would spend an hour and address this.

It was more like 5 hours of work  and 4 days that I'm waiting to merge it :) but, right  I think now we understand it better. The route takes a long time to be processed. To memoize code which is used in a loop, relevant just for an instance, is a valid to pattern. The cache query doesn't invalidate it, but helps. If that's the best pattern to implement cache in controllers, it's a different question.. anyway so now I see three options:

a) drop the PR, keep the memoization as it is
b) accept the PR, keep the memoization at it is but with a better code
c) propose a better solution for caching in this controller, implement it, measure and see if it worth it.

I'm for b) and you?

Comment (coolo) [reply src/api/app/controllers/status_project_controller.rb]: > we should have a look and verify that we need it 
That's exactly what I asked for - and nothing more.

When I say "be careful" does it sound to you like "don't do it"? If so, I'm sorry for the confusion

**Comment by hennevogel** [comment]
If you just want to dance around this, sure...
